### PR TITLE
feat(pipeline executions/orca): Add support for max concurrent pipeline executions

### PR DESCRIPTION
**Pull request by guido9j**
It adds support for max concurrent pipeline executions. If concurrent pipeline execution is enabled, pipelines will queue when the max concurrent pipeline executions is reached. Any queued pipelines will be allowed to run once the number of running pipeline executions drops below the max. If the max is set to 0, then pipelines will not queue.

This is the proposed change for https://github.com/spinnaker/spinnaker/issues/6558.

Changes to deck, front50 and echo are also needed in order to expose max concurrent pipeline executions as a pipeline configuration to Spinnaker users.

(Front50 pull request: spinnaker/front50#1080)
(Echo pull request: https://github.com/spinnaker/echo/pull/1134)
(Deck pull request: https://github.com/spinnaker/deck/pull/9777)

### CODE REVIEW

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (mattgogerly) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:124]: What's this magic 2?

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (mattgogerly) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: Why less than 0?

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java
Comment (mattgogerly) [inline orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java:864]: Does this need to be supported for SQL backed Orca also?

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (guido9j) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:124]: I added 2, because of the code that checks for any **RUNNING** pipeline a few lines below starts with a page size of 2. Plus 1 may work, but to be in the safe side I added 2:
`val criteria = ExecutionCriteria().setPageSize(2).setStatuses(RUNNING)`

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (guido9j) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: If `isLimitConcurrent == false && maxConcurrentExecutions <= 0` then don't queue. This implements/keeps the current logic, where `isLimitConcurrent == false` then don't queue.

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java
Comment (guido9j) [reply orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java:864]: I could not find any changes needed for the SQL backed Orca.

**Comment by guido9j** [comment]
@mattgogerly: I replied on your comments. Let me know what you think. Thanks

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (mattgogerly) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:124]: Makes sense, might make sense to extract it out to a constant so we don't end up peppering numbers throughout

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (mattgogerly) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: Shouldn't it just return false always, if limitConcurrent is false? Why is the check for maxConcurrent necessary?

Comment (mattgogerly) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: Shouldn't it just return false always, if limitConcurrent is false? Why is the check for maxConcurrent necessary?

Comment (guido9j) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: It needs to return true if `isLimitConcurrent == false && maxConcurrentExecutions > 0 && RUNNING >= maxConcurrentExecutions`. If makes more readable, I can change to:

      !isLimitConcurrent -> {
        return when {
          getMaxConcurrentExecutions() > 0 -> {
            val criteria = ExecutionCriteria().setPageSize(maxConcurrentExecutions+2).setStatuses(RUNNING)
            repository
              .retrievePipelinesForPipelineConfigId(configId, criteria)
              .filter { it.id != id }
              .count()
              .toBlocking()
              .first() >= getMaxConcurrentExecutions()
          }
          else -> false
        }
      }

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (guido9j) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:124]: I'll change the `2` into a constant set to 2. I can name it 'MIN_PAGE_SIZE`.

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt
Comment (mattgogerly) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/OrcaMessageHandler.kt:122]: Ah my bad, I had the implication of limitConcurrent as "have an upper limit on concurrent" (which is obviously what you're implementing) rather than "prevent concurrent altogether".

Sorry!

**Comment by mattgogerly** [comment]
Looks good to me, but I'll defer to somebody a bit more experienced.

@dbyron-sf @german-muzquiz

**Comment by guido9j** [comment]
@mattgogerly , @german-muzquiz : I looked at the build failure over the past few days, and as far as I can tell this is caused by a commit to master on Dec 4th. The failure is intermittent. I ran the build on the master branch, and it failed 3 out of 5 times:
Screen Shot 2021-12-07 at 7 25 03 AM

**Comment by guido9j** [comment]
Here is the output from test that is failing now:
[TEST-com.netflix.spinnaker.orca.controllers.AdminControllerTest.xml.log](https://github.com/spinnaker/orca/files/7668471/TEST-com.netflix.spinnaker.orca.controllers.AdminControllerTest.xml.log)

**Comment by guido9j** [comment]
Last build on master failed with same error: https://github.com/spinnaker/orca/runs/4438645255?check_suite_focus=true

**Comment by dbyron-sf** [comment]
> Last build on master failed with same error: https://github.com/spinnaker/orca/runs/4438645255?check_suite_focus=true

Yup, it's a flaky test...would love some help tracking down the issue.  In the meantime, my magic bullet is re-running failed jobs...

**Comment by dbyron-sf** [comment]
@mergifyio update

**Comment by guido9j** [comment]
@dbyron-sf, @mattgogerly, @german-muzquiz : can this be merged to master? Thanks

**Review by link108 — APPROVED** [review]

**Comment by link108** [comment]
@guido9j Approved! Could you please update the branch and/or make it editable by maintainers? eg https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

Then I can update and merge the branch 👍

**Comment by guido9j** [comment]
@link108: The option "Allow edit from maintainers" is not available for me, since this is an organization-owned fork (not sure if I just don't have the needed permissions in my org). Either way, I went ahead and merged the latest changes from master and build is in progress now. Thanks

**Comment by link108** [comment]
Thank you!

**Pull request merged**
